hs-bindgen/src-internal/HsBindgen.hs

```diff
@@ -73,14 +74,14 @@ hsBindgen
       -- 2. Frontend.
       frontendArtefact <-
         frontend tracerFrontend bindgenFrontendConfig bootArtefact
-      pure (bootArtefact, frontendArtefact)
-    (bootArtefact, frontendArtefact) <- either throwIO pure eArtefact
-    -- 3. Backend.
-    backendArtefact <- withTracerSafe tracerConfigSafe $ \tracer -> do
-      backend tracer bindgenBackendConfig bootArtefact frontendArtefact
-    -- 4. Artefacts.
-    withTracerSafe tracerConfigSafe $ \tracer -> do
-      runArtefacts tracer bootArtefact frontendArtefact backendArtefact artefacts
+      -- 3. Backend.
+      backendArtefact <- withTracerSafe tracerConfigSafe $ \tracer' -> do
+        backend tracer' bindgenBackendConfig bootArtefact frontendArtefact
+      -- 4. Artefacts.
+      withTracerSafe tracerConfigSafe $ \tracer' -> do
```

Comment:
Yes, I agree. It is suboptimal that we have the safe trace intertwined here. This makes me think we should probably combine the two. Edsko and I thought about this quite a lot, and the beauty of the current approach is that the backend truly only can omit safe traces. On the other hand, I am pretty sure we can lift safe traces into the other tracers at this stage too.